File: java/singleton.java
/** 饿汉式 */
public class Singleton {
  private static Singleton instance = new Singleton();
  private Singleton() {}

  public static Singleton getInstance() {
    return instance;
  }

  public void greet() {
    System.out.println("Hello ~");
  }
}

public class SingletonPatternDemo {
  public static void main(String[] args) {
    Singleton s = Singleton.getInstance();
    s.greet();
  }
}

File: js/singleton.js
function Singleton() {
  this.greet = function() {
    console.log('Hello ~');
  };
}

Singleton.getInstance = (function() {
  var instance;

  return function() {
    if (instance) {
      return instance;
    }

    return (instance = new Singleton());
  }
})();

var o1 = Singleton.getInstance();
var o2 = Singleton.getInstance();

console.log(o1 === o2);